### ISSUE TITLE
[Ask] Is File Cache Support Tagging?

### ISSUE THREAD

**Issue by imannms** [swayok/alternative-laravel-cache#3]

**Comment by swayok** [comment]
Yes. Also it supports hierarchial cache (http://www.php-cache.com/en/latest/hierarchy/)

**Comment by imannms** [comment]
How to use hierarchial cache with Laravel?

I have tried:

```php
        $user = 4711;
        for ($i = 0; $i < 100; $i++) {
            $key = sprintf('|users|%d|followers|%d|likes', $user, $i);
            Cache::tags(['test', 'hierarchy'])->put($key, 'Justin Bieber', 10);
        }

        Cache::has('|users|4711|followers|12|likes'); // true

        Cache::forget('|users|4711|followers');

        Cache::has('|users|4711|followers|12|likes'); // true

```
but not working

**Comment by swayok** [comment]
What is your cache driver config in config/cache.php?

**Comment by imannms** [comment]
```php
<?php

return [

    'default' => env('CACHE_DRIVER', 'file'),

    'stores' => [

        'apc' => [
            'driver' => 'apc',
        ],

        'array' => [
            'driver' => 'array',
        ],

        'database' => [
            'driver' => 'database',
            'table' => 'cache',
            'connection' => null,
        ],

        'file' => [
            'driver' => 'file',
            'path' => storage_path('framework/cache/data'),
        ],

        'memcached' => [
            'driver' => 'memcached',
            'persistent_id' => env('MEMCACHED_PERSISTENT_ID'),
            'sasl' => [
                env('MEMCACHED_USERNAME'),
                env('MEMCACHED_PASSWORD'),
            ],
            'options' => [
                // Memcached::OPT_CONNECT_TIMEOUT  => 2000,
            ],
            'servers' => [
                [
                    'host' => env('MEMCACHED_HOST', '127.0.0.1'),
                    'port' => env('MEMCACHED_PORT', 11211),
                    'weight' => 100,
                ],
            ],
        ],

        'redis' => [
            'driver' => 'redis',
            'connection' => 'default',
        ],

    ],

    'prefix' => env(
        'CACHE_PREFIX',
        str_slug(env('APP_NAME', 'laravel'), '_').'_cache'
    ),

];
```

**Comment by swayok** [comment]
Can you replace in `src/AlternativeLaravelCache/Store/AlternativeHierarchialFileCacheStore.php`
method `fixItemKey` with next code and try again? Maybe last change in this method made hierarchial cache unavailable.

```
public function fixItemKey($key) {
        // allowed chars: "a-zA-Z0-9_\.! |"
        return parent::fixItemKey(preg_replace(
            ['%-+%',   '%/+%', '%[^a-zA-Z0-9_\.! |]+%'],
            ['_dash_', '_ds_', '_'],
            $key
        ));
    }
```

**Comment by imannms** [comment]
Yes. It's working now.

**Comment by swayok** [comment]
Ok. I will rollback changes to this file then.

**Comment by swayok** [comment]
Fixed in version 5.4.3.
Make sure you specify new driver name for hierarchial file cache: `hierarchial_file`

**Comment by imannms** [comment]
updated to version 5.4.3

Error:
Cache store [hierarchial_file] is not defined.
